### PR TITLE
:books: Add design for field write semantics

### PR DESCRIPTION
**Pull request by elbeno**

### CODE REVIEW

File: docs/field_write_semantics.adoc [hunk 1]
```diff
@@ -0,0 +1,129 @@
+
+== Field Write Semantics
+
+Each field in a register is parameterized with a position -- the closed
+(inclusive) range from LSB to MSB, from which a mask can be derived -- and a
+_write function_.
```
Comment (lukevalenty) [inline docs/field_write_semantics.adoc:6]: Suggestion: I think "write function" is the most accurate mathematical term for this...but something like "access modifier" may be a more accessible term for users.

You could go on to explain that the access modifiers are modeled as write functions and carry on with the explanation below.

File: docs/field_write_semantics.adoc [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+^|X ^|X ^|X
+
+^|* ^|* ^|*
+|===
+
+=== Building the masks
+
+In general a given write to a register will only want to write some of the field
+values. The other field values are left unspecified. We can compute the common
+identity spec for the write by folding the spec function across all the fields
```
Comment (lukevalenty) [inline docs/field_write_semantics.adoc:92]: Question: do we need to find a common identity spec across all unspecified fields? i believe we can evaluate the identity function for each field on its own to determine if a RMW is necessary or we can just write 1 or 0.

Comment (elbeno) [reply docs/field_write_semantics.adoc:92]: I think we're saying the same thing. The identity spec function _is_ how we find out whether RMW is necessary. The identity for a field isn't a function, it's a value. This is very general: it allows for fields in a register to have different write functions.

File: docs/field_write_semantics.adoc (hunk 1 above)
Comment (elbeno) [reply docs/field_write_semantics.adoc:6]: Sure - at the moment I just wanted to write down this theory. I reality users don't really need to know about this stuff; we  can totally present the options as just a few of the conventionally-named possible values (eg read_only).

File: docs/field_write_semantics.adoc (hunk 2 above)
Comment (elbeno) [reply docs/field_write_semantics.adoc:92]: Bear in mind that "all the fields across the register" could also mean "all the fields across the addressable unit" if we combine this with byte enables.

**Pull request merged**